### PR TITLE
[BugFix] fix core dump on DotRewriter pass

### PR DESCRIPTION
**Pull request by Yancey1989**
This PR fixed core dump while two `dot_general` ops have the same producer `transpose` op, the following Canonicalizer pass would remove the transpose op if no consumer.

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/dot_rewriter.cc [hunk 1]
```diff
@@ -395,15 +395,6 @@ struct TransposeFoldingConvert : public OpRewritePattern<mhlo::DotGeneralOp> {
     Value dot = rewriter.create<mhlo::DotGeneralOp>(
         loc, op.getType(), lhs, rhs, dot_dimension_attr, nullptr);
     rewriter.replaceOp(op, dot);
-
-    // Remove transpose op which outputs into dot.
-    if (tp_lhs) {
-      rewriter.eraseOp(old_lhs.getDefiningOp());
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/dot_rewriter.cc:401]: Maybe we can add a filecheck based test? For example:

```
%0 = transpose // unsafe to remove transpose directly
%1 = matmul(%0, ...)
return %0, %1
```

Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/dot_rewriter.cc:401]: done

File: tao_compiler/mlir/disc/transforms/tests/dot_rewriter.mlir [hunk 2]
```diff
@@ -131,3 +131,15 @@ func.func @dot_general_1dx1d(%arg0: tensor<256xf32>, %arg1: tensor<256xf32>) ->
...
+// -----
+// CHECK-LABEL: func.func @dot_general_transpose(
+// CHECK-SAME:        %[[ARG0:.*]]: tensor<?x?xf32>, %[[ARG1:.*]]: tensor<?xf32>) -> tensor<?x?xf32> {
+// CHECK:         %[[T0:.*]] = "mhlo.dot_general"(%[[ARG0]], %[[ARG1]]) {dot_dimension_numbers = #mhlo.dot<lhs_contracting_dimensions = [0], rhs_contracting_dimensions = [0]>} : (tensor<?x?xf32>, tensor<?xf32>) -> tensor<?x?xf32>
+// CHECK:         return %[[T0]]
+// CHECK:       }
+func.func @dot_general_transpose(%arg0: tensor<?x?xf32>, %arg1: tensor<?xf32>) -> tensor<?x?xf32> {
+  %1 = "mhlo.transpose"(%arg0) {permutation = dense<[1, 0]> : tensor<2xi64>} : (tensor<?x?xf32>) -> tensor<?x?xf32>
+  %2 = "mhlo.dot_general"(%1, %arg1) {dot_dimension_numbers = #mhlo.dot<lhs_contracting_dimensions = [1], rhs_contracting_dimensions = [0]>} : (tensor<?x?xf32>, tensor<?xf32>) -> tensor<?x?xf32>
+  return %2 : tensor<?x?xf32>
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/tests/dot_rewriter.mlir:144]: nit: maybe we need to return `%1` as well to guard the change of this PR

**Review by wyzero — APPROVED** [review]

File: tao_compiler/mlir/disc/transforms/tests/dot_rewriter.mlir (hunk 2 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/tests/dot_rewriter.mlir:144]: thanks, will update in the next pr.

**Pull request merged**